### ISSUE TITLE
Twitter DMs returning unparsable Unicode

### ISSUE THREAD

**Issue by tibbon** [sferik/twitter#849]
Environment: Twitter gem v5.15.0. Rails 4.2.8, Ruby 2.3.2. 

I understand there's a chance this is patched under a more recent version, but I haven't been able to find an issue or PR related.

In a request for retrieving DMs, I'm getting a parsing error that appears to be from Twitter returning potentially invalid Unicode. Of course, ideally Twitter would never return invalid unicode, but we don't control them. 

In the return body of a DMs retrieval request (I can't post the entire thing, because it belongs to a client). The vast majority (99.99%) of DM retrieval works in my codebase, so it isn't a code error otherwise on my end.

```ruby
"text":"Hello! \ud83d\ude2d\ud83d... "
```

It appears to be an issue from the `Twitter:REST:Response#parse` call to `JSON.parse(body, :symbolize_names => true)`. I'm calling the `twitter_rest_client_instance.direct_messages(since: last_message_id, count: 200)` method. Below is my stack trace: 

```ruby
JSON::ParserError: invalid escaped character at line 1, column 20187 [parse.c:259]
  from twitter/rest/response/parse_json.rb:15:in `parse'
  from twitter/rest/response/parse_json.rb:15:in `parse'
  from twitter/rest/response/parse_json.rb:20:in `on_complete'
  from faraday/response.rb:9:in `block in call'
  from faraday/response.rb:57:in `on_complete'
  from faraday/response.rb:8:in `call'
  from faraday/response.rb:8:in `call'
  from faraday/request/url_encoded.rb:15:in `call'
  from faraday/request/multipart.rb:14:in `call'
  from twitter/rest/request/multipart_with_file.rb:19:in `call'
  from faraday/rack_builder.rb:139:in `build_response'
  from faraday/connection.rb:377:in `run_request'
  from faraday/connection.rb:140:in `get'
  from twitter/rest/request.rb:33:in `perform'
  from twitter/rest/utils.rb:50:in `perform_request'
  from twitter/rest/utils.rb:95:in `perform_request_with_objects'
  from twitter/rest/utils.rb:80:in `perform_get_with_objects'
  from twitter/rest/direct_messages.rb:27:in `direct_messages_received'
  from twitter/rest/direct_messages.rb:90:in `direct_messages'
  from lib/twitter_api_service.rb:66:in `block in direct_messages'
  from lib/hyper/utils.rb:23:in `social_network_request'
  from lib/twitter_api_service.rb:65:in `direct_messages'
  from app/workers/twitter_direct_message_retrieval_worker.rb:8:in `perform'
  from sidekiq/processor.rb:158:in `execute_job'
  from sidekiq/processor.rb:138:in `block (4 levels) in process'
  from sidekiq.rb:36:in `block in <module:Sidekiq>'
  from sidekiq/processor.rb:133:in `block (3 levels) in process'
  from sidekiq/middleware/chain.rb:128:in `block in invoke'
  from new_relic/agent/instrumentation/sidekiq.rb:33:in `block in call'
  from new_relic/agent/instrumentation/controller_instrumentation.rb:363:in `perform_action_with_newrelic_trace'
  from new_relic/agent/instrumentation/sidekiq.rb:29:in `call'
  from sidekiq/middleware/chain.rb:130:in `block in invoke'
  from sidekiq/middleware/server/active_record.rb:6:in `call'
  from sidekiq/middleware/chain.rb:130:in `block in invoke'
  from raven/integrations/sidekiq.rb:7:in `call'
  from sidekiq/middleware/chain.rb:130:in `block in invoke'
  from sidekiq/batch/middleware.rb:33:in `call'
  from sidekiq/middleware/chain.rb:130:in `block in invoke'
  from sidekiq-ent/limiter/middleware.rb:38:in `call'
  from sidekiq/middleware/chain.rb:130:in `block in invoke'
  from sidekiq/middleware/server/retry_jobs.rb:74:in `call'
  from sidekiq/middleware/chain.rb:130:in `block in invoke'
  from sidekiq/failures/middleware.rb:9:in `call'
  from sidekiq/middleware/chain.rb:130:in `block in invoke'
  from sidekiq/middleware/server/logging.rb:11:in `block in call'
  from sidekiq/logging.rb:32:in `with_context'
  from sidekiq/middleware/server/logging.rb:7:in `call'
  from sidekiq/middleware/chain.rb:130:in `block in invoke'
  from sidekiq/middleware/chain.rb:133:in `invoke'
  from sidekiq/processor.rb:132:in `block (2 levels) in process'
  from sidekiq/processor.rb:174:in `stats'
  from sidekiq/processor.rb:131:in `block in process'
  from sidekiq.rb:35:in `block in <module:Sidekiq>'
  from sidekiq/processor.rb:126:in `process'
  from sidekiq/processor.rb:82:in `process_one'
  from sidekiq/processor.rb:70:in `run'
  from sidekiq/util.rb:17:in `watchdog'
  from sidekiq/util.rb:26:in `block in safe_thread'
Twitter::Error: invalid escaped character at line 1, column 20187 [parse.c:259]
  from twitter/rest/request.rb:37:in `rescue in perform'
  from twitter/rest/request.rb:32:in `perform'
  from twitter/rest/utils.rb:50:in `perform_request'
  from twitter/rest/utils.rb:95:in `perform_request_with_objects'
  from twitter/rest/utils.rb:80:in `perform_get_with_objects'
  from twitter/rest/direct_messages.rb:27:in `direct_messages_received'
  from twitter/rest/direct_messages.rb:90:in `direct_messages'
  from lib/twitter_api_service.rb:66:in `block in direct_messages'
```

I can attempt to write a failing test for it. Since it deals with something pretty deep inside Ruby's JSON parsing, I'm not sure what we can do about it though.

**Comment by tibbon** [comment]
There's now a $100 bug bounty on this: 

https://www.bountysource.com/issues/45140196-twitter-dms-returning-unparsable-unicode

**Comment by stve** [comment]
hey @tibbon, 

JSON parsing is ultimately left to the `json` library so it seems that this issue is most likely upstream. I see that you opened an issue on `flori/json` but it doesn't look like it has been addressed yet. 

I think the best approach is to create a failing test there and see if that can get things headed toward resolution. I'm not a C programmer, but I did attempt to locate the code that raises the error only to find that `parse.c` does not exist in the code base. 😕
